flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/BipartiteGraph.java

```diff
@@ -0,0 +1,272 @@
+ * a bipartite graph should be first converted into a {@link Graph} instance. This can be achieved by using
+ * {@link BipartiteGraph#simpleTopProjection()} or
+ * {@link BipartiteGraph#fullBottomProjection()} methods.
+ *
+ * @param <TK> the key type of the top vertices
+ * @param <BK> the key type of the bottom vertices
+ * @param <TV> the top vertices value type
+ * @param <BV> the bottom vertices value type
+ * @param <EV> the edge value type
+ */
+public class BipartiteGraph<TK, BK, TV, BV, EV> {
+	private final ExecutionEnvironment context;
+	private final DataSet<Vertex<TK, TV>> topVertices;
+	private final DataSet<Vertex<BK, BV>> bottomVertices;
+	private final DataSet<BipartiteEdge<TK, BK, EV>> edges;
+
+	private BipartiteGraph(
+			DataSet<Vertex<TK, TV>> topVertices,
+			DataSet<Vertex<BK, BV>> bottomVertices,
+			DataSet<BipartiteEdge<TK, BK, EV>> edges,
+			ExecutionEnvironment context) {
+		this.topVertices = topVertices;
+		this.bottomVertices = bottomVertices;
+		this.edges = edges;
+		this.context = context;
+	}
+
+	/**
+	 * Create bipartite graph from datasets.
+	 *
+	 * @param topVertices  dataset of top vertices in the graph
+	 * @param bottomVertices dataset of bottom vertices in the graph
+	 * @param edges dataset of edges between vertices
+	 * @param context Flink execution context
+	 * @param <KT> the key type of the top vertices
+	 * @param <KB> the key type of the bottom vertices
+	 * @param <VT> the top vertices value type
+	 * @param <VB> the bottom vertices value type
+	 * @param <EV> the edge value type
+	 * @return new bipartite graph created from provided datasets
+	 */
+	public static <KT, KB, VT, VB, EV> BipartiteGraph<KT, KB, VT, VB, EV> fromDataSet(
+			DataSet<Vertex<KT, VT>> topVertices,
+			DataSet<Vertex<KB, VB>> bottomVertices,
+			DataSet<BipartiteEdge<KT, KB, EV>> edges,
+			ExecutionEnvironment context) {
+		return new BipartiteGraph<>(topVertices, bottomVertices, edges, context);
+	}
+
+	/**
+	 * Get dataset with top vertices.
+	 *
+	 * @return dataset with top vertices
+	 */
+	public DataSet<Vertex<TK, TV>> getTopVertices() {
+		return topVertices;
+	}
+
+	/**
+	 * Get dataset with bottom vertices.
+	 *
+	 * @return dataset with bottom vertices
+	 */
+	public DataSet<Vertex<BK, BV>> getBottomVertices() {
+		return bottomVertices;
+	}
+
+	/**
+	 * Get dataset with graph edges.
+	 *
+	 * @return dataset with graph edges
+	 */
+	public DataSet<BipartiteEdge<TK, BK, EV>> getEdges() {
+		return edges;
+	}
+
+	/**
+	 * Convert a bipartite into a graph that contains only top vertices. An edge between two vertices in the new
+	 * graph will exist only if the original bipartite graph contains a bottom vertex they are both connected to.
+	 *
+	 * @return top projection of the bipartite graph where every edge contains a tuple with values of two edges that
+	 * connect top vertices in the original graph
+	 */
+	public Graph<TK, TV, Tuple2<EV, EV>> simpleTopProjection() {
```

Comment:
Also, and it was discussed only have a Tuple2 of edge values, but double checking that we don't also want to include the (here: bottom) vertex ID in the new edge value.